File: src/services/writing_helper.py
# -*- coding: utf-8 -*-
"""Safe writing-helper capabilities for Niko Studio."""

from __future__ import annotations

from typing import Any, Dict
import re

_ALLOWED_MODES = {"polish", "summarize", "outline", "rewrite", "expand"}
_SENTENCE_ENDINGS = ".!?。！？"
_CONCISE_HINTS = ("简洁", "精简", "更短", "concise", "shorter")
_FORMAL_HINTS = ("正式", "书面", "formal")
_CASUAL_HINTS = ("口语", "轻松", "casual")


def _normalize_whitespace(text: str) -> str:
    normalized = text.replace("\r\n", "\n").replace("\r", "\n")
    normalized = re.sub(r"[\t\f\v]+", " ", normalized)
    normalized = re.sub(r"[ ]{2,}", " ", normalized)
    normalized = re.sub(r"\n{3,}", "\n\n", normalized)
    return normalized.strip()


def _split_sentences(text: str) -> list[str]:
    matches = re.findall(r".+?[。！？.!?]|.+$", text)
    return [part.strip() for part in matches if part and part.strip()]


def _summarize_text(content: str, max_sentences: int) -> str:
    sentences = _split_sentences(content)
    if not sentences:
        return ""
    selected = sentences[: max(1, max_sentences)]
    return " ".join(selected).strip()


def _outline_text(content: str, max_items: int) -> list[str]:
    paragraphs = [item.strip() for item in content.split("\n") if item.strip()]
    if not paragraphs:
        return []

    outline: list[str] = []
    for paragraph in paragraphs:
        lead = _summarize_text(paragraph, 1)
        if lead:
            outline.append(lead)
        if len(outline) >= max(1, max_items):
            break
    return outline


def _normalize_punctuation_spacing(text: str) -> str:
    normalized = re.sub(r"[ \t]+([，。！？；：,.!?;:])", r"\1", text)
    normalized = re.sub(r"([,.;:!?])([^\s])", r"\1 \2", normalized)
    normalized = re.sub(r"([，。！？；：])[ \t]+", r"\1", normalized)
    return normalized


def _compress_redundant_phrases(text: str) -> str:
    compressed = text
    replacements = (
        (r"\b(in order to)\b", "to"),
        (r"\b(due to the fact that)\b", "because"),
        (r"为了能够", "为"),
        (r"为了可以", "为"),
        (r"非常非常", "非常"),
        (r"很很", "很"),
    )
    for pattern, replacement in replacements:
        compressed = re.sub(pattern, replacement, compressed, flags=re.IGNORECASE)
    compressed = re.sub(r"\b(\w+)\s+\1\b", r"\1", compressed, flags=re.IGNORECASE)
    return compressed


def _sentence_identity(sentence: str) -> str:
    compact = re.sub(r"\s+", "", sentence).lower()
    return compact.strip(_SENTENCE_ENDINGS)


def _dedupe_adjacent_sentences(text: str) -> str:
    lines = text.split("\n")
    deduped_lines: list[str] = []

    for line in lines:
        stripped_line = line.strip()
        if not stripped_line:
            deduped_lines.append("")
            continue

        sentences = _split_sentences(stripped_line)
        if not sentences:
            deduped_lines.append(stripped_line)
            continue

        deduped: list[str] = []
        previous_identity = ""
        for sentence in sentences:
            identity = _sentence_identity(sentence)
            if identity and identity == previous_identity:
                continue
            deduped.append(sentence)
            previous_identity = identity

        deduped_lines.append(" ".join(deduped).strip())

    merged = "\n".join(deduped_lines)
    merged = re.sub(r"\n{3,}", "\n\n", merged)
    return merged.strip()


def _apply_instruction_tone(text: str, instruction: str) -> str:
    normalized_instruction = _normalize_whitespace(instruction).lower()
    if not normalized_instruction:
        return text

    result = text

    concise_enabled = any(hint in normalized_instruction for hint in _CONCISE_HINTS)
    formal_enabled = any(hint in normalized_instruction for hint in _FORMAL_HINTS)
    casual_enabled = any(hint in normalized_instruction for hint in _CASUAL_HINTS)

    if concise_enabled:
        fillers = (
            r"\bactually\b",
            r"\bbasically\b",
            r"\bin fact\b",
            "其实",
            "基本上",
            "某种程度上",
        )
        for filler in fillers:
            result = re.sub(filler, "", result, flags=re.IGNORECASE)
        result = _compress_redundant_phrases(result)

    if formal_enabled and not casual_enabled:
        formal_map = {
            "can't": "cannot",
            "don't": "do not",
            "it's": "it is",
            "挺": "较为",
            "有点": "略微",
            "真的": "确实",
        }
        for src, target in formal_map.items():
            result = result.replace(src, target)

    if casual_enabled and not formal_enabled:
        casual_map = {
            "therefore": "so",
            "furthermore": "also",
            "因此": "所以",
            "此外": "还有",
            "较为": "挺",
            "略微": "有点",
        }
        for src, target in casual_map.items():
            result = result.replace(src, target)

    return _normalize_whitespace(result)


def _polish_text(content: str, instruction: str = "") -> str:
    polished = _normalize_whitespace(content)
    polished = _normalize_punctuation_spacing(polished)
    polished = _compress_redundant_phrases(polished)
    polished = _dedupe_adjacent_sentences(polished)
    polished = _apply_instruction_tone(polished, instruction)
    return _normalize_whitespace(polished)


def _rewrite_text(content: str, instruction: str = "") -> str:
    rewritten = _polish_text(content, instruction=instruction)
    sentences = _split_sentences(rewritten)
    if len(sentences) >= 2:
        rewritten = " ".join(sentences)
    return _normalize_whitespace(rewritten)


def _expand_text(content: str, instruction: str = "") -> str:
    expanded_base = _polish_text(content, instruction=instruction)
    sentences = _split_sentences(expanded_base)
    if not sentences:
        return expanded_base

    anchor = sentences[0]
    if anchor and anchor[-1] in _SENTENCE_ENDINGS:
        addition = f"进一步展开：{anchor}"
    else:
        addition = f"进一步展开：{anchor}。"
    return _normalize_whitespace(f"{expanded_base}\n\n{addition}")


def process_writing_helper(
    *,
    content: str,
    mode: str = "polish",
    max_sentences: int = 3,
    max_items: int = 6,
    instruction: str = "",
) -> Dict[str, Any]:
    normalized_content = _normalize_whitespace(content)
    normalized_mode = str(mode or "polish").strip().lower()
    normalized_instruction = instruction if isinstance(instruction, str) else ""

    if normalized_mode not in _ALLOWED_MODES:
        raise ValueError("mode must be one of: polish, summarize, outline, rewrite, expand")

    if normalized_mode == "polish":
        polished = _polish_text(normalized_content, instruction=normalized_instruction)
        return {
            "mode": "polish",
            "processed_text": polished,
            "stats": {
                "input_chars": len(content),
                "output_chars": len(polished),
            },
        }

    if normalized_mode == "rewrite":
        rewritten = _rewrite_text(normalized_content, instruction=normalized_instruction)
        return {
            "mode": "rewrite",
            "processed_text": rewritten,
            "stats": {
                "input_chars": len(content),
                "output_chars": len(rewritten),
            },
        }

    if normalized_mode == "expand":
        expanded = _expand_text(normalized_content, instruction=normalized_instruction)
        return {
            "mode": "expand",
            "processed_text": expanded,
            "stats": {
                "input_chars": len(content),
                "output_chars": len(expanded),
            },
        }

    if normalized_mode == "summarize":
        summary = _summarize_text(normalized_content, max_sentences=max_sentences)
        return {
            "mode": "summarize",
            "processed_text": summary,
            "stats": {
                "input_chars": len(content),
                "output_chars": len(summary),
                "max_sentences": max(1, max_sentences),
            },
        }

    outline = _outline_text(normalized_content, max_items=max_items)
    return {
        "mode": "outline",
        "outline": outline,
        "stats": {
            "input_chars": len(content),
            "items": len(outline),
            "max_items": max(1, max_items),
        },
    }

File: src-ts/services/writing-helper.ts
/**
 * WritingHelper - Safe writing-helper capabilities for Niko Studio
 *
 * Migrated from src/services/writing_helper.py.
 *
 * Provides text processing modes: polish, summarize, outline, rewrite, expand.
 */

const _ALLOWED_MODES = new Set(['polish', 'summarize', 'outline', 'rewrite', 'expand']);
const _SENTENCE_ENDINGS = '.!?。！？';
const _CONCISE_HINTS = ['简洁', '精简', '更短', 'concise', 'shorter'];
const _FORMAL_HINTS = ['正式', '书面', 'formal'];
const _CASUAL_HINTS = ['口语', '轻松', 'casual'];

/**
 * Normalize whitespace in text
 */
function _normalizeWhitespace(text: string): string {
  let normalized = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
  normalized = normalized.replace(/[\t\f\v]+/g, ' ');
  normalized = normalized.replace(/[ ]{2,}/g, ' ');
  normalized = normalized.replace(/\n{3,}/g, '\n\n');
  return normalized.trim();
}

/**
 * Split text into sentences (supports Chinese and English punctuation)
 */
function _splitSentences(text: string): string[] {
  const matches = text.match(/.+?[。！？.!?]|.+$/g);
  return matches
    ? matches.map((part) => part.trim()).filter((part) => part.length > 0)
    : [];
}

/**
 * Summarize text to a maximum number of sentences
 */
function _summarizeText(content: string, maxSentences: number): string {
  const sentences = _splitSentences(content);
  if (sentences.length === 0) {
    return '';
  }
  const selected = sentences.slice(0, Math.max(1, maxSentences));
  return selected.join(' ').trim();
}

/**
 * Generate an outline from text
 */
function _outlineText(content: string, maxItems: number): string[] {
  const paragraphs = content
    .split('\n')
    .map((item) => item.trim())
    .filter((item) => item.length > 0);
  if (paragraphs.length === 0) {
    return [];
  }

  const outline: string[] = [];
  for (const paragraph of paragraphs) {
    const lead = _summarizeText(paragraph, 1);
    if (lead) {
      outline.push(lead);
    }
    if (outline.length >= Math.max(1, maxItems)) {
      break;
    }
  }
  return outline;
}

/**
 * Normalize punctuation spacing
 */
function _normalizePunctuationSpacing(text: string): string {
  let normalized = text.replace(/[ \t]+([，。！？；：,.!?;:])/g, '$1');
  normalized = normalized.replace(/([,.;:!?])([^\s])/g, '$1 $2');
  normalized = normalized.replace(/([，。！？；：])[ \t]+/g, '$1');
  return normalized;
}

/**
 * Compress redundant phrases in text
 */
function _compressRedundantPhrases(text: string): string {
  let compressed = text;
  const replacements: [RegExp, string][] = [
    [/\b(in order to)\b/gi, 'to'],
    [/\b(due to the fact that)\b/gi, 'because'],
    [/为了能够/g, '为'],
    [/为了可以/g, '为'],
    [/非常非常/g, '非常'],
    [/很很/g, '很'],
  ];
  for (const [pattern, replacement] of replacements) {
    compressed = compressed.replace(pattern, replacement);
  }
  compressed = compressed.replace(/\b(\w+)\s+\1\b/gi, '$1');
  return compressed;
}

/**
 * Compute sentence identity for deduplication
 */
function _sentenceIdentity(sentence: string): string {
  const compact = sentence.replace(/\s+/g, '').toLowerCase();
  return compact.replace(new RegExp(`[${_escapeRegex(_SENTENCE_ENDINGS)}]+$`), '');
}

/**
 * Escape string for use in RegExp
 */
function _escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Deduplicate adjacent sentences within each line
 */
function _dedupeAdjacentSentences(text: string): string {
  const lines = text.split('\n');
  const dedupedLines: string[] = [];

  for (const line of lines) {
    const strippedLine = line.trim();
    if (!strippedLine) {
      dedupedLines.push('');
      continue;
    }

    const sentences = _splitSentences(strippedLine);

    const deduped: string[] = [];
    let previousIdentity = '';
    for (const sentence of sentences) {
      const identity = _sentenceIdentity(sentence);
      if (identity && identity === previousIdentity) {
        continue;
      }
      deduped.push(sentence);
      previousIdentity = identity;
    }

    dedupedLines.push(deduped.join(' ').trim());
  }

  let merged = dedupedLines.join('\n');
  merged = merged.replace(/\n{3,}/g, '\n\n');
  return merged.trim();
}

/**
 * Apply instruction tone hints (concise, formal, casual)
 */
function _applyInstructionTone(text: string, instruction: string): string {
  const normalizedInstruction = _normalizeWhitespace(instruction).toLowerCase();
  if (!normalizedInstruction) {
    return text;
  }

  let result = text;

  const conciseEnabled = _CONCISE_HINTS.some((hint) => normalizedInstruction.includes(hint));
  const formalEnabled = _FORMAL_HINTS.some((hint) => normalizedInstruction.includes(hint));
  const casualEnabled = _CASUAL_HINTS.some((hint) => normalizedInstruction.includes(hint));

  if (conciseEnabled) {
    const fillers = [
      /\bactually\b/gi,
      /\bbasically\b/gi,
      /\bin fact\b/gi,
      /其实/g,
      /基本上/g,
      /某种程度上/g,
    ];
    for (const filler of fillers) {
      result = result.replace(filler, '');
    }
    result = _compressRedundantPhrases(result);
  }

  if (formalEnabled && !casualEnabled) {
    const formalMap: Record<string, string> = {
      "can't": 'cannot',
      "don't": 'do not',
      "it's": 'it is',
      '挺': '较为',
      '有点': '略微',
      '真的': '确实',
    };
    for (const [src, target] of Object.entries(formalMap)) {
      result = result.replaceAll(src, target);
    }
  }

  if (casualEnabled && !formalEnabled) {
    const casualMap: Record<string, string> = {
      'therefore': 'so',
      'furthermore': 'also',
      '因此': '所以',
      '此外': '还有',
      '较为': '挺',
      '略微': '有点',
    };
    for (const [src, target] of Object.entries(casualMap)) {
      result = result.replaceAll(src, target);
    }
  }

  return _normalizeWhitespace(result);
}

/**
 * Check if text has adjacent duplicate sentences
 */
function _hasAdjacentDuplicateSentence(text: string): boolean {
  for (const line of text.split('\n')) {
    const strippedLine = line.trim();
    if (!strippedLine) {
      continue;
    }
    const sentences = _splitSentences(strippedLine);
    let previousIdentity = '';
    for (const sentence of sentences) {
      const identity = _sentenceIdentity(sentence);
      if (identity && identity === previousIdentity) {
        return true;
      }
      previousIdentity = identity;
    }
  }
  return false;
}

/**
 * Polish text: normalize, deduplicate, apply tone
 */
function _polishText(content: string, instruction: string = ''): string {
  let polished = _normalizeWhitespace(content);
  polished = _normalizePunctuationSpacing(polished);
  polished = _compressRedundantPhrases(polished);
  const hadAdjacentDuplicates = _hasAdjacentDuplicateSentence(polished);
  polished = _dedupeAdjacentSentences(polished);
  polished = _applyInstructionTone(polished, instruction);
  polished = _normalizeWhitespace(polished);
  if (hadAdjacentDuplicates) {
    polished = polished
      .split('\n')
      .map((line) => line.trim())
      .filter((line) => line.length > 0)
      .join(' ');
  }
  return _normalizeWhitespace(polished);
}

/**
 * Rewrite text: polish then rejoin sentences
 */
function _rewriteText(content: string, instruction: string = ''): string {
  let rewritten = _polishText(content, instruction);
  const sentences = _splitSentences(rewritten);
  if (sentences.length >= 2) {
    rewritten = sentences.join(' ');
  }
  return _normalizeWhitespace(rewritten);
}

/**
 * Expand text: polish then add expansion hint
 */
function _expandText(content: string, instruction: string = ''): string {
  const expandedBase = _polishText(content, instruction);
  const sentences = _splitSentences(expandedBase);
  if (sentences.length === 0) {
    return expandedBase;
  }

  const anchor = sentences[0];
  const endsWithSentenceEnding = anchor.length > 0 && _SENTENCE_ENDINGS.includes(anchor[anchor.length - 1]);
  const addition = endsWithSentenceEnding
    ? `进一步展开：${anchor}`
    : `进一步展开：${anchor}。`;
  return _normalizeWhitespace(`${expandedBase}\n\n${addition}`);
}

/**
 * Process writing helper request
 *
 * @param params - Processing parameters
 * @param params.content - Input text content
 * @param params.mode - Processing mode: polish, summarize, outline, rewrite, expand
 * @param params.maxSentences - Max sentences for summarize mode (default 3)
 * @param params.maxItems - Max items for outline mode (default 6)
 * @param params.instruction - Tone/style instruction hints
 * @returns Processing result with mode-specific output and stats
 */
export function processWritingHelper(params: {
  content: string;
  mode?: string;
  maxSentences?: number;
  maxItems?: number;
  instruction?: string;
}): Record<string, unknown> {
  const {
    content,
    mode = 'polish',
    maxSentences = 3,
    maxItems = 6,
    instruction = '',
  } = params;

  const normalizedContent = _normalizeWhitespace(content);
  const normalizedMode = (mode || 'polish').trim().toLowerCase();
  const normalizedInstruction = typeof instruction === 'string' ? instruction : '';

  if (!_ALLOWED_MODES.has(normalizedMode)) {
    throw new Error('mode must be one of: polish, summarize, outline, rewrite, expand');
  }

  if (normalizedMode === 'polish') {
    const polished = _polishText(normalizedContent, normalizedInstruction);
    return {
      mode: 'polish',
      processed_text: polished,
      stats: {
        input_chars: content.length,
        output_chars: polished.length,
      },
    };
  }

  if (normalizedMode === 'rewrite') {
    const rewritten = _rewriteText(normalizedContent, normalizedInstruction);
    return {
      mode: 'rewrite',
      processed_text: rewritten,
      stats: {
        input_chars: content.length,
        output_chars: rewritten.length,
      },
    };
  }

  if (normalizedMode === 'expand') {
    const expanded = _expandText(normalizedContent, normalizedInstruction);
    return {
      mode: 'expand',
      processed_text: expanded,
      stats: {
        input_chars: content.length,
        output_chars: expanded.length,
      },
    };
  }

  if (normalizedMode === 'summarize') {
    const summary = _summarizeText(normalizedContent, maxSentences);
    return {
      mode: 'summarize',
      processed_text: summary,
      stats: {
        input_chars: content.length,
        output_chars: summary.length,
        max_sentences: Math.max(1, maxSentences),
      },
    };
  }

  // outline mode
  const outline = _outlineText(normalizedContent, maxItems);
  return {
    mode: 'outline',
    outline,
    stats: {
      input_chars: content.length,
      items: outline.length,
      max_items: Math.max(1, maxItems),
    },
  };
}
